### PR TITLE
Abstract Model API further

### LINKED ISSUE #37: Fix bug where num_warmup isn't properly saved to a Model object
https://github.com/gibsramen/BIRDMAn/blob/e7bd19b7aee6ea2c7010b75bebcc2c13c3b4c012/birdman/model_base.py#L61-L62

If `num_warmup` is passed explicitly, no `self.num_warmup` attribute is created which causes issues in `_fit_serial` and `_fit_parallel`.

### PR DESCRIPTION
**Pull request by gibsramen**
cc @mortonjt

Relevant to #34, closes #37

Creates a new class `BaseModel` that doesn't require a formula for a design matrix. `Model` class renamed to `RegressionModel`.

### CODE REVIEW

File: birdman/model_base.py [hunk 1]
```diff
@@ -60,26 +56,21 @@ def __init__(
         self.num_iter = num_iter
         if num_warmup is None:
             self.num_warmup = num_iter
+        else:
+            self.num_warmup = num_warmup
```
Comment (mortonjt) [inline birdman/model_base.py:60]: 👍

File: birdman/default_models.py [hunk 2]
```diff
@@ -16,11 +16,12 @@
         "features": os.path.join(TEMPLATES, "negative_binomial_single.stan"),
         "lme": os.path.join(TEMPLATES, "negative_binomial_lme.stan")
     },
-    "multinomial": os.path.join(TEMPLATES, "multinomial.stan")
+    "multinomial": os.path.join(TEMPLATES, "multinomial.stan"),
+    "pln": os.path.join(TEMPLATES, "poisson_lognormal.stan")
```
Comment (mortonjt) [inline birdman/default_models.py:20]: Looks like this was inherited from another pull request - just don't forget about addressing this.

**Review by mortonjt — APPROVED** [review]
This is great! This more or less addresses my issues with the base model before.

We'll need to readdress the dask load balancing issue at some point - I think the scaling of the cluster may have to be moved to outside of the base class.  But that'll require a little tinkering that may be best to save for another pull request.

File: birdman/default_models.py (hunk 2 above)
Comment (gibsramen) [reply birdman/default_models.py:20]: Good catch, thanks. This is a vestigial change from the PLN branch that slipped through.

**Comment by gibsramen** [comment]
Thanks! Yeah we can work on the load balancing stuff separately.

**Pull request merged**